pandas/core/generic.py

```diff
@@ -489,16 +489,51 @@ def _get_index_resolvers(self):
             d.update(self._get_axis_resolvers(axis_name))
         return d
 
-    def _get_space_character_free_column_resolvers(self):
-        """Return the space character free column resolvers of a dataframe.
+    def _get_special_character_free_column_resolvers(self, expr):
+        """Return the special character free column resolvers of a dataframe.
 
-        Column names with spaces are 'cleaned up' so that they can be referred
-        to by backtick quoting.
+        Column names with special characters are 'cleaned up' so that they can
+        be referred to by backtick quoting.
         Used in :meth:`DataFrame.eval`.
         """
-        from pandas.core.computation.common import _remove_spaces_column_name
+        from pandas.core.computation.expr import (
+            _BACKTICK_QUOTED_STRING,
+            _clean_special_characters_column_name,
+            _tokenize_string,
+        )
+
+        resolvers = [
```

Comment:
Hmm so what is the point of converting `self.items()` to a list of tuples to ultimately convert back to a dict? IIUC you are just trying to detect if the cleansed names clash and get duplicated right? If so why not keep this a dict and just compare the length to `self.items()` after the fact?

Reply:
I need to have `resolvers` as a list of tuples as I am actually wanting to detect the exact name that causes such a clash. 

I do this in the next code line by creating `cleaned_name_count`, a Counter object containing counts for the cleansed names. If there is a cleansed name with count > 1, we know it clashed an I throw an error. 

Is this a problem with performance? If so, I can do as you recommend and only start the procedure to find the clash when the dicts have different length.

Reply:
Not a performance issue just readability. Wouldn't taking the set difference of the dict keys before and after give you the same information?

Reply:
No, the cleansing changes the keys of the dictionary. The values stay the same though, but hashing the values would really be a performance issue. 

After thinking about it, I do think your proposed method is a more readable approach. It really looks a lot more like a guard clause that way. 

But I don't think there is nice way to do the check elegantly, and get the clashed names. You just have to redo the conversion.

Also, another complicated factor. I only throw the error if the expr uses a name from one of the columns that clashes. So, if the dataframe has clashing column names, but the expr does not uses one of them. It is not a problem and the error is not raised. 

Okay, I am gonna try something, and you can say if this is an improvement or not.